File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CcdAuthenticator.java
package uk.gov.hmcts.reform.bulkscan.orchestrator.services.ccd;

import java.util.function.Supplier;

public class CcdAuthenticator {

    private final String userId;
    private final Supplier<String> serviceTokenSupplier;
    private final String userToken;

    public CcdAuthenticator(
        Supplier<String> serviceTokenSupplier,
        String userId,
        String userToken
    ) {
        this.serviceTokenSupplier = serviceTokenSupplier;
        this.userId = userId;
        this.userToken = userToken;
    }

    public String getUserToken() {
        return this.userToken;
    }

    public String getServiceToken() {
        return this.serviceTokenSupplier.get();
    }

    public String getUserId() {
        return this.userId;
    }
}

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CcdAuthenticator.kt
package uk.gov.hmcts.reform.bulkscan.orchestrator.services.ccd

import uk.gov.hmcts.reform.idam.client.models.UserDetails
import java.util.function.Supplier
import kotlin.properties.ReadOnlyProperty
import kotlin.reflect.KProperty

/**
 * To support the easy interop with the java code supplier to the delegated property
 */
fun Supplier<String>.asReadOnlyProp(): ReadOnlyProperty<CcdAuthenticator, String> {
    return object : ReadOnlyProperty<CcdAuthenticator, String> {
        override fun getValue(thisRef: CcdAuthenticator, property: KProperty<*>): String {
            return this@asReadOnlyProp.get()
        }
    }
}

class CcdAuthenticator(
    serviceTokenSupplier: ReadOnlyProperty<CcdAuthenticator, String>,
    val userDetails: UserDetails,
    userTokenSupplier: ReadOnlyProperty<CcdAuthenticator, String>
) {

    val userToken: String by userTokenSupplier
    val serviceToken: String by serviceTokenSupplier

    companion object {
        @JvmStatic
        fun from(
            serviceTokenSupplier: Supplier<String>,
            userDetails: UserDetails,
            userTokenSupplier: Supplier<String>
        ): CcdAuthenticator {
            return CcdAuthenticator(
                serviceTokenSupplier.asReadOnlyProp(),
                userDetails,
                userTokenSupplier.asReadOnlyProp())
        }
    }
}
